### PR TITLE
Feature/dont email unverified

### PR DESCRIPTION
**Pull request by Baggerone**
Don't send emails when an unverified Mfa option is deleted

### CODE REVIEW

File: application/common/components/Emailer.php [hunk 1]
```diff
@@ -493,30 +493,32 @@ public function shouldSendMfaEnabledMessageTo($user, $mfaEventType)
     }
 
     /**
-     * Whether the user just deleted an MFA option and there is still one or more left
+     * Whether the user just deleted a verified MFA option and there is still one or more left
      *
      * @param User $user (assumes the user instance has already been refreshed)
      * @param string Mfa::EVENT_TYPE_*
      * @return bool
```
Comment (forevermatt) [inline application/common/components/Emailer.php]: Looks like you'll also want to add an `@param` on each of these functions for the new `$mfa` parameter.

File: application/features/email.feature [hunk 2]
```diff
@@ -185,7 +185,6 @@ Feature: Email
       | to send                 | does exist     | does NOT exist  | delete_mfa        | should      |
       | to send                 | does NOT exist | does exist      | delete_mfa        | should      |
       | to send                 | does exist     | does exist      | delete_mfa        | should      |
-      | to send                 | does NOT exist | does NOT exist  | delete_mfa        | should NOT  |
```
Comment (forevermatt) [inline application/features/email.feature]: Out of curiosity, why was this example removed from the test?

File: application/features/email.feature [hunk 3]
```diff
@@ -205,11 +204,38 @@ Feature: Email
       | to send            | does exist     | does NOT exist  | delete_mfa   | should NOT  |
       | to send            | does NOT exist | does exist      | delete_mfa   | should NOT  |
       | to send            | does exist     | does exist      | delete_mfa   | should NOT  |
-      | to send            | does NOT exist | does NOT exist  | delete_mfa   | should      |
```
Comment (forevermatt) [inline application/features/email.feature]: Same here (I'm curious why this test case was removed).

File: application/features/email.feature [hunk 4]
```diff
@@ -205,11 +204,38 @@ Feature: Email
...
       | to send            | does NOT exist | does NOT exist  | verify_mfa   | should NOT  |
       | NOT to send        | does NOT exist | does NOT exist  | verify_mfa   | should NOT  |
 
+
+  Scenario Outline: When to send an email after a verified u2f mfa option has been deleted.
+    Given we are configured to send mfa disabled emails
+      And we are configured to send mfa option removed emails
+      And no mfas exist
+      And a user already exists
+      And a verified u2f mfa option used to exist
```
Comment (forevermatt) [inline application/features/email.feature]: What does it mean that a verified U2F MFA option "used to exist"?

File: application/features/email.feature [hunk 5]
```diff
@@ -205,11 +204,38 @@ Feature: Email
...
+      And the latest mfa event type was delete_mfa
+    When I check if a mfa <optionRemovedOrDisabled> email should be sent
+    Then I see that a mfa <optionRemovedOrDisabled> email <shouldOrNot> be sent
+
+    Examples:
+      | optionRemovedOrDisabled | shouldOrNot |
+      | option removed          | should NOT  |
```
Comment (forevermatt) [inline application/features/email.feature]: I think there's a typo here: when we have just deleted a verified MFA option, then we _should_ send the "MFA option removed" email. Right?

File: application/features/email.feature [hunk 6]
```diff
@@ -205,11 +204,38 @@ Feature: Email
...
+      | disabled                | should      |
+
+  Scenario: When to send an email after an unverified u2f mfa option has been deleted.
+    Given we are configured to send mfa disabled emails
```
Comment (forevermatt) [inline application/features/email.feature]: Whether we are configured to send MFA-disabled emails seems irrelevant to this test (since the When/Then only care about MFA-option-removed emails), so this line could be removed and the `Given` moved down to the next line.

File: application/features/email.feature [hunk 7]
```diff
@@ -205,11 +204,38 @@ Feature: Email
...
+      And we are configured to send mfa option removed emails
+      And no mfas exist
+      And a user already exists
+      And an unverified u2f mfa option does exist
+      And a verified u2f mfa option does exist but not as the test mfa
+      And the latest mfa event type was delete_mfa
+    When I check if a mfa option removed email should be sent
```
Comment (forevermatt) [inline application/features/email.feature]: Based on the wording of this test scenario, I don't think it's describing the situation we're trying to test here. (For example, it doesn't seem to say whether the MFA option just deleted was verified or not... simply that there is an unverified MFA option and a verified MFA option that do exist).

What if we changed from these lines...
```gherkin
    And an unverified u2f mfa option does exist
    And a verified u2f mfa option does exist but not as the test mfa
    And the latest mfa event type was delete_mfa
```
... to this:
```gherkin
    And a verified u2f mfa option does exist
    And an unverified u2f mfa option was just deleted
```

File: application/features/email.feature (hunk 2 above)
Comment (Baggerone) [reply application/features/email.feature]: It was moved down to a separate test, because this test doesn't have a way of saying that  the Mfa used to exist but doesn't anymore.  It just says that right now, it doesn't exist.  So, it didn't have access to a Mfa instance to send to the Emailer method.

File: application/features/email.feature (hunk 4 above)
Comment (Baggerone) [reply application/features/email.feature]: It gives the test a way to hold on to a Mfa instance to send to the Emailer class. It doesn't save it to the database and then delete it, since deleting it isn't doable with the external service we use.

File: application/features/email.feature (hunk 5 above)
Comment (Baggerone) [reply application/features/email.feature]: When it's the last mfa option of that type, we only send the mfa disabled email and not the mfa option removed email.

Comment (forevermatt) [reply application/features/email.feature]: I don't think we differentiate by MFA type when deciding whether to sent the "MFA option removed" vs. the "MFA disabled" emails: if they have _any_ remaining MFA options, then they have simply removed an MFA option; if they don't have any MFA options remaining, then they have disabled MFA.

Comment (forevermatt) [reply application/features/email.feature]: Never mind, you're right. I was overlooking that this was testing which email to send when the _last_ verified MFA option is deleted.

File: application/features/email.feature [hunk 8]
```diff
@@ -205,11 +204,35 @@ Feature: Email
...
+      And a verified u2f mfa option was just deleted
+    When I check if a mfa <optionRemovedOrDisabled> email should be sent
+    Then I see that a mfa <optionRemovedOrDisabled> email <shouldOrNot> be sent
+
+    Examples:
+      | optionRemovedOrDisabled | shouldOrNot |
+      | option removed          | should NOT  |
+      | disabled                | should      |
+
+  Scenario: When to send an email after an unverified u2f mfa option has been deleted.
```
Comment (forevermatt) [inline application/features/email.feature]: This test scenario looks fine now. Should we also have one like this but for testing the sending of the MFA-disabled email?
